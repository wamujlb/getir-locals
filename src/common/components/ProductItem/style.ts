import { Theme } from '@mui/material';
import { createStyles, makeStyles } from '@mui/styles';

export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      display: 'flex',
      flexDirection: 'column',
      height: '100%',
      justifyContent: 'space-between',
      '& + $root': {
        marginTop: theme.spacing(3),
      },
    },
    title: {
      color: theme.palette.text.secondary,
      marginBottom: theme.spacing(1.5),
      fontWeight: theme.typography.fontWeightMedium,
    },
    price: {
      color: theme.palette.text.secondary,
      marginBottom: theme.spacing(1.5),
      fontWeight: theme.typography.fontWeightMedium,
    },
  })
);
