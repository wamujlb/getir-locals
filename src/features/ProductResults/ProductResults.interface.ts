export interface ProductItemData {
  readonly id: string;
  readonly title: string;
  readonly price: number;
  readonly imageUri: string;
}
