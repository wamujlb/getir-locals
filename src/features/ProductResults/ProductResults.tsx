import React, { useEffect, useState } from 'react';
import { useSelector } from 'react-redux';
import {
  Card,
  CardContent,
  Pagination,
  PaginationRenderItemParams,
} from '@mui/material';

import ProductResult from './ProductResult/ProductResult';
import { selectShoppingCartItems } from 'store/ducks/shoppingCart/shoppingCart.slice';
import { ProductsResults } from 'api/productsApi/productsApi.interface';
import { useAppDispatch } from 'store/useAppDispatch';
import PaginationNext from 'common/components/Pagination/PaginationNext/PaginationNext';
import PaginationPrev from 'common/components/Pagination/PaginationPrev/PaginationPrev';
import StyledPaginationItem from 'common/components/Pagination/StyledPaginationItem/StyledPaginationItem';
import ProductItemSkeleton from 'common/components/ProductItem/ProductItemSkeleton/ProductItemSkeleton';
import {
  getProducts,
  selectProductsData,
  selectProductsError,
  selectProductsFailure,
  selectProductsFetching,
  selectProductsLoading,
} from 'store/ducks/products/products.slice';
import ProductResultsContent from './ProductResultsContent/ProductResultsContent';
import { DEFAULT_PAGE_SIZE } from 'api/productsApi/productsApi';

const ProductResults = () => {
  const dispatch = useAppDispatch();
  const shoppingCart = useSelector(selectShoppingCartItems);
  const products = useSelector(selectProductsData);
  const isLoaiding = useSelector(selectProductsLoading);
  const isFailure = useSelector(selectProductsFailure);
  const isFetching = useSelector(selectProductsFetching);
  const error = useSelector(selectProductsError);
  const [page, setPage] = useState<number>(1);

  useEffect(() => {
    dispatch(getProducts({ page }));
  }, [dispatch, page]);

  const handlePageChange = (_: React.ChangeEvent<unknown>, newPage: number) => {
    setPage(newPage);
  };

  return (
    <>
      <Card>
        <CardContent>
          <ProductResultsContent
            pageSize={DEFAULT_PAGE_SIZE}
            isLoading={isLoaiding}
            isFailure={isFailure}
            isFetching={isFetching}
            error={error}
            shoppingCart={shoppingCart}
            data={products}
          />
        </CardContent>
      </Card>

      {products.count > 0 && (
        <Pagination
          count={products.count}
          shape="rounded"
          sx={{ display: 'flex', justifyContent: 'center', mt: 4 }}
          onChange={handlePageChange}
          renderItem={(item: PaginationRenderItemParams) => (
            <StyledPaginationItem
              {...item}
              components={{ previous: PaginationPrev, next: PaginationNext }}
              color="primary"
            />
          )}
        />
      )}
    </>
  );
};

ProductResults.displayName = 'ProductResults';

export default ProductResults;
