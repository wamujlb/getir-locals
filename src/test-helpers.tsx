import React from 'react';
import { ThemeProvider } from '@mui/styles';

import { theme } from 'theme/theme';

export const WrappedWithTheme = ({ children }: any) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};
