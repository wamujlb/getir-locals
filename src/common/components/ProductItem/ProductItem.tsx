import React from 'react';
import { Card, CardContent, CardProps, Typography } from '@mui/material';

import { useStyles } from './style';
import { usePrice } from 'common/hooks/usePrice';
import { ProductItemData } from 'features/ProductResults/ProductResults.interface';
import FetchButton from '../Buttons/FetchButton/FetchButton';
import ProductImage from '../ProductImage/ProductImage';

export type ProductItemProps = {
  addedToCart?: boolean;
  onAddToCart: (e: React.MouseEvent<HTMLButtonElement>) => void;
  onRemoveFromCart: (e: React.MouseEvent<HTMLButtonElement>) => void;
} & ProductItemData &
  CardProps;

const ProductItem = React.memo(
  ({
    title,
    price,
    addedToCart,
    imageUri,
    onAddToCart,
    onRemoveFromCart,
  }: ProductItemProps) => {
    const classes = useStyles();
    const formattedPrice = usePrice({ value: price });

    const handleClick = (e: React.MouseEvent<HTMLButtonElement>) => {
      if (addedToCart) {
        onRemoveFromCart(e);
        return;
      }

      onAddToCart(e);
    };

    return (
      <Card className={classes.root}>
        <div>
          <ProductImage productTitle="title" src={imageUri} />

          <Typography
            variant="body2"
            fontWeight="bold"
            sx={{ color: 'primary.main', mt: 1 }}
          >
            {formattedPrice}
          </Typography>

          <Typography variant="body2" sx={{ color: 'text.primary', mb: 0.5 }}>
            {title}
          </Typography>
        </div>

        <FetchButton
          fullWidth
          size="small"
          color="primary"
          variant={addedToCart ? 'outlined' : 'contained'}
          onClick={handleClick}
        >
          {addedToCart ? 'Remove' : 'Add'}
        </FetchButton>
      </Card>
    );
  }
);

ProductItem.displayName = 'ProductItem';

export default ProductItem;
