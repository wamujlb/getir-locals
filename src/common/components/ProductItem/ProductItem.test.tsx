import React from 'react';
import { render, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import ShoppingCartIcon from 'common/components/icons/ShoppingCart/ShoppingCart';

import ProductItem, { ProductItemProps } from './ProductItem';
import { ProductItemData } from 'features/ProductResults/ProductResults.interface';
import { WrappedWithTheme } from 'test-helpers';

const data: ProductItemData = {
  id: '1',
  title: 'Gorgeous Office Mug',
  price: 14.99,
  imageUri: 'https://picsum.photos/300',
};

const TestComponent = (props: ProductItemProps) => {
  return (
    <WrappedWithTheme>
      <ProductItem {...props} />
    </WrappedWithTheme>
  );
};

describe('#ProductItem', () => {
  const onAddToCartFn = jest.fn();
  const onRemoveFromCartFn = jest.fn();

  afterAll(() => {
    onAddToCartFn.mockClear();
    onRemoveFromCartFn.mockClear();
    cleanup();
  });

  it('should match the snapshot when product not added to the cart', async () => {
    const screen = render(
      <TestComponent
        onAddToCart={onAddToCartFn}
        onRemoveFromCart={onRemoveFromCartFn}
        {...data}
      />
    );

    expect(screen.container).toMatchSnapshot();
  });

  it('should match the snapshot when product is added to the cart', async () => {
    const screen = render(
      <TestComponent
        onAddToCart={onAddToCartFn}
        onRemoveFromCart={onRemoveFromCartFn}
        addedToCart
        {...data}
      />
    );

    expect(screen.container).toMatchSnapshot();
  });

  it('should call expected functions when depending on addedToCart prop', async () => {
    const screen = render(
      <TestComponent
        onAddToCart={onAddToCartFn}
        onRemoveFromCart={onRemoveFromCartFn}
        {...data}
      />
    );

    userEvent.click(screen.getByText('Add'));

    expect(onAddToCartFn).toHaveBeenCalled();

    screen.rerender(
      <TestComponent
        addedToCart
        onAddToCart={onAddToCartFn}
        onRemoveFromCart={onRemoveFromCartFn}
        {...data}
      />
    );

    userEvent.click(screen.getByText('Remove'));

    expect(onRemoveFromCartFn).toHaveBeenCalled();
  });
});
