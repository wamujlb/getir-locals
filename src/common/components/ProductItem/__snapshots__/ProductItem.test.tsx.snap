// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#ProductItem should match the snapshot when product is added to the cart 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root makeStyles-root-6 css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div>
      <div
        class="makeStyles-root-9"
      >
        <img
          alt="title"
          class="makeStyles-image-10"
          src="https://picsum.photos/300"
        />
      </div>
      <p
        class="MuiTypography-root MuiTypography-body2 css-kl203v-MuiTypography-root"
      >
        ₺ 14.99
      </p>
      <p
        class="MuiTypography-root MuiTypography-body2 css-clx6gs-MuiTypography-root"
      >
        Gorgeous Office Mug
      </p>
    </div>
    <button
      class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-fullWidth MuiButtonBase-root  css-1jb5gvc-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Remove
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`#ProductItem should match the snapshot when product not added to the cart 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root makeStyles-root-1 css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div>
      <div
        class="makeStyles-root-4"
      >
        <img
          alt="title"
          class="makeStyles-image-5"
          src="https://picsum.photos/300"
        />
      </div>
      <p
        class="MuiTypography-root MuiTypography-body2 css-kl203v-MuiTypography-root"
      >
        ₺ 14.99
      </p>
      <p
        class="MuiTypography-root MuiTypography-body2 css-clx6gs-MuiTypography-root"
      >
        Gorgeous Office Mug
      </p>
    </div>
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-fullWidth MuiButtonBase-root  css-1evprfk-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Add
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
